File: src-tauri/src/services/permission.service.rs
use serde_json::{json, Value};

use crate::entities::permission_entity::{TodoPermission, ASSIGNEE_DEFAULT_ROLE};

#[derive(Debug, Clone)]
pub struct PermissionService {}

impl PermissionService {
  pub fn get_todo_permission(todo: &Value, user_id: &str) -> Option<TodoPermission> {
    Self::get_todo_permission_with_profile(todo, user_id, None)
  }

  pub fn get_todo_permission_with_profile(
    todo: &Value,
    user_id: &str,
    profile_id: Option<&str>,
  ) -> Option<TodoPermission> {
    Self::get_todo_permission_with_profile_and_admin(todo, user_id, profile_id, false)
  }

  pub fn get_todo_permission_with_profile_and_admin(
    todo: &Value,
    user_id: &str,
    profile_id: Option<&str>,
    is_global_admin: bool,
  ) -> Option<TodoPermission> {
    let owner_id = todo.get("user_id").and_then(|v| v.as_str()).unwrap_or("");

    if owner_id == user_id {
      return Some(TodoPermission::OWNER);
    }

    let assignees = todo.get("assignees").and_then(|v| v.as_array())?;
    let assignee_roles = todo.get("assignee_roles").and_then(|v| v.as_object());

    for assignee_id in assignees {
      let assignee_str = assignee_id.as_str()?;

      let role = assignee_roles
        .and_then(|r| r.get(assignee_str))
        .and_then(|v| v.as_str())
        .unwrap_or(ASSIGNEE_DEFAULT_ROLE);

      if assignee_str == profile_id.unwrap_or("") || assignee_str == user_id {
        return Some(TodoPermission::from_str(role));
      }
    }

    let visibility = todo
      .get("visibility")
      .and_then(|v| v.as_str())
      .unwrap_or("private");
    match visibility {
      "public" => {
        if is_global_admin {
          Some(TodoPermission::MODERATOR)
        } else {
          Some(TodoPermission::VIEWER)
        }
      }
      "shared" => {
        if is_global_admin {
          return Some(TodoPermission::MODERATOR);
        }
        if let Some(pid) = profile_id {
          if assignees.iter().any(|a| a.as_str() == Some(pid)) {
            return Some(TodoPermission::VIEWER);
          }
        }
        if assignees.iter().any(|a| a.as_str() == Some(user_id)) {
          Some(TodoPermission::VIEWER)
        } else {
          None
        }
      }
      _ => None,
    }
  }

  pub fn is_owner_or_admin(todo: &Value, user_id: &str) -> bool {
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      return permission == TodoPermission::MODERATOR || permission == TodoPermission::OWNER;
    }
    false
  }

  pub fn can_view_todo(todo: &Value, user_id: &str) -> bool {
    let visibility = todo
      .get("visibility")
      .and_then(|v| v.as_str())
      .unwrap_or("private");
    let owner_id = todo.get("user_id").and_then(|v| v.as_str()).unwrap_or("");
    let assignees = todo.get("assignees").and_then(|v| v.as_array());

    match visibility {
      "private" => owner_id == user_id,
      "shared" => {
        owner_id == user_id
          || assignees
            .map(|a| a.iter().any(|id| id.as_str() == Some(user_id)))
            .unwrap_or(false)
      }
      "public" => true,
      _ => false,
    }
  }

  pub fn can_edit_todo(todo: &Value, user_id: &str) -> bool {
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      return permission.can_edit_todo_fields();
    }
    false
  }

  #[allow(dead_code)]
  pub fn can_manage_assignees(todo: &Value, user_id: &str) -> bool {
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      return permission == TodoPermission::OWNER;
    }
    false
  }

  pub fn can_delete_todo(todo: &Value, user_id: &str) -> bool {
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      return permission.can_delete_todo();
    }
    false
  }

  pub fn can_add_task_to_todo(todo: &Value, user_id: &str) -> bool {
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      return permission.can_create_task();
    }
    Self::can_view_todo(todo, user_id)
  }

  pub fn can_edit_task(task: &Value, todo: &Value, user_id: &str) -> bool {
    if Self::is_owner_or_admin(todo, user_id) {
      return true;
    }
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      if permission.can_edit_task() {
        let task_creator_id = task.get("user_id").and_then(|v| v.as_str()).unwrap_or("");
        return task_creator_id == user_id;
      }
    }
    false
  }

  pub fn can_delete_task(task: &Value, todo: &Value, user_id: &str) -> bool {
    if Self::is_owner_or_admin(todo, user_id) {
      return true;
    }
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      if permission.can_delete_task() {
        let task_creator_id = task.get("user_id").and_then(|v| v.as_str()).unwrap_or("");
        return task_creator_id == user_id;
      }
    }
    false
  }

  pub fn can_edit_subtask(subtask: &Value, _task: &Value, todo: &Value, user_id: &str) -> bool {
    if Self::is_owner_or_admin(todo, user_id) {
      return true;
    }
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      if permission.can_edit_subtask() {
        let subtask_creator_id = subtask
          .get("user_id")
          .and_then(|v| v.as_str())
          .unwrap_or("");
        return subtask_creator_id == user_id;
      }
    }
    false
  }

  pub fn can_delete_subtask(subtask: &Value, _task: &Value, todo: &Value, user_id: &str) -> bool {
    if Self::is_owner_or_admin(todo, user_id) {
      return true;
    }
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      if permission.can_delete_subtask() {
        let subtask_creator_id = subtask
          .get("user_id")
          .and_then(|v| v.as_str())
          .unwrap_or("");
        return subtask_creator_id == user_id;
      }
    }
    false
  }

  #[allow(dead_code)]
  pub fn can_edit_comment(comment: &Value, _task: &Value, todo: &Value, user_id: &str) -> bool {
    if Self::is_owner_or_admin(todo, user_id) {
      return true;
    }
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      if permission.can_edit_comment() {
        let comment_creator_id = comment
          .get("user_id")
          .and_then(|v| v.as_str())
          .unwrap_or("");
        return comment_creator_id == user_id;
      }
    }
    false
  }

  #[allow(dead_code)]
  pub fn can_delete_comment(comment: &Value, _task: &Value, todo: &Value, user_id: &str) -> bool {
    if Self::is_owner_or_admin(todo, user_id) {
      return true;
    }
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      if permission.can_delete_comment() {
        let comment_creator_id = comment
          .get("user_id")
          .and_then(|v| v.as_str())
          .unwrap_or("");
        return comment_creator_id == user_id;
      }
    }
    false
  }

  #[allow(dead_code)]
  pub fn can_archive_task(task: &Value, todo: &Value, user_id: &str) -> bool {
    if Self::is_owner_or_admin(todo, user_id) {
      return true;
    }
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      if permission.can_archive_task() {
        let task_creator_id = task.get("user_id").and_then(|v| v.as_str()).unwrap_or("");
        return task_creator_id == user_id;
      }
    }
    false
  }

  #[allow(dead_code)]
  pub fn can_archive_subtask(subtask: &Value, _task: &Value, todo: &Value, user_id: &str) -> bool {
    if Self::is_owner_or_admin(todo, user_id) {
      return true;
    }
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      if permission.can_archive_subtask() {
        let subtask_creator_id = subtask
          .get("user_id")
          .and_then(|v| v.as_str())
          .unwrap_or("");
        return subtask_creator_id == user_id;
      }
    }
    false
  }

  #[allow(dead_code)]
  pub fn can_archive_comment(comment: &Value, _task: &Value, todo: &Value, user_id: &str) -> bool {
    if Self::is_owner_or_admin(todo, user_id) {
      return true;
    }
    if let Some(permission) = Self::get_todo_permission(todo, user_id) {
      if permission.can_archive_comment() {
        let comment_creator_id = comment
          .get("user_id")
          .and_then(|v| v.as_str())
          .unwrap_or("");
        return comment_creator_id == user_id;
      }
    }
    false
  }

  pub fn get_todo_filter_for_user(
    user_id: &str,
    profile_id: Option<&str>,
    visibility: Option<&str>,
  ) -> Value {
    match visibility.unwrap_or("private") {
      "private" => {
        json!({
            "visibility": "private",
            "user_id": user_id
        })
      }
      "shared" => {
        let assignee_filter = if let Some(pid) = profile_id {
          json!({ "assignees": { "$in": [pid, user_id] } })
        } else {
          json!({ "assignees": { "$in": [user_id] } })
        };
        json!({
            "visibility": "shared",
            "$or": [
                { "user_id": user_id },
                assignee_filter
            ]
        })
      }
      "public" => {
        json!({ "visibility": "public" })
      }
      "all" => {
        let shared_assignee_filter = if let Some(pid) = profile_id {
          json!({ "assignees": { "$in": [pid, user_id] } })
        } else {
          json!({ "assignees": { "$in": [user_id] } })
        };
        json!({
            "$or": [
                { "visibility": "private", "user_id": user_id },
                { "visibility": "shared", "$or": [{ "user_id": user_id }, shared_assignee_filter] },
                { "visibility": "public" }
            ]
        })
      }
      _ => {
        json!({ "visibility": "private", "user_id": user_id })
      }
    }
  }

  pub fn can_view_category(category: &Value, user_id: &str) -> bool {
    let visibility = category
      .get("visibility")
      .and_then(|v| v.as_str())
      .unwrap_or("private");
    let owner_id = category
      .get("user_id")
      .and_then(|v| v.as_str())
      .unwrap_or("");

    match visibility {
      "private" => owner_id == user_id,
      "shared" => owner_id == user_id,
      "public" => true,
      _ => false,
    }
  }

  pub fn can_edit_category(category: &Value, user_id: &str) -> bool {
    let owner_id = category
      .get("user_id")
      .and_then(|v| v.as_str())
      .unwrap_or("");
    owner_id == user_id
  }

  pub fn can_delete_category(category: &Value, user_id: &str) -> bool {
    Self::can_edit_category(category, user_id)
  }

  pub fn get_category_filter_for_user(user_id: &str, visibility: Option<&str>) -> Value {
    match visibility.unwrap_or("private") {
      "private" => {
        json!({ "user_id": user_id })
      }
      "shared" => {
        json!({
            "$or": [
                { "user_id": user_id },
                { "visibility": "shared" }
            ]
        })
      }
      "public" => {
        json!({
            "$or": [
                { "user_id": user_id },
                { "visibility": "shared" },
                { "visibility": "public" }
            ]
        })
      }
      _ => {
        json!({ "user_id": user_id })
      }
    }
  }
}

File: src/app/core/services/permission.service.ts
import { Injectable, inject, signal } from "@angular/core";
import { firstValueFrom } from "rxjs";
import { Todo, Task, Subtask, Comment } from "@entities/generated/api.types";
import { ApiService } from "@services/api.service";
import { JwtTokenService } from "@services/auth/jwt-token.service";
import { TodoPermissionLevel, PermissionCheckResult } from "@entities/entity-config.model";
export enum TodoPermission {
  VIEWER = "viewer",
  EDITOR = "editor",
  MODERATOR = "moderator",
  OWNER = "owner",
}
export interface TodoPermissionContext {
  todo: Todo;
  userId: string;
  assigneeRoles: Record<string, string>;
  effectivePermission: TodoPermission;
}
export interface TaskPermissionContext {
  task: Task;
  todo: Todo;
  userId: string;
  todoPermission: TodoPermission;
}
export interface SubtaskPermissionContext {
  subtask: Subtask;
  task: Task;
  todo: Todo;
  userId: string;
  todoPermission: TodoPermission;
}
@Injectable({ providedIn: "root" })
export class PermissionService {
  private api = inject(ApiService);
  private jwtTokenService = inject(JwtTokenService);
  private permissionCache = signal<Map<string, TodoPermission>>(new Map());
  isGlobalAdmin(): boolean {
    const token = this.jwtTokenService.getToken();
    return this.jwtTokenService.hasRole(token, "admin");
  }
  getCurrentUserId(): string {
    return this.jwtTokenService.getCurrentUserId() || "";
  }
  fromStr(role: string): TodoPermission {
    switch (role.toLowerCase()) {
      case "viewer":
        return TodoPermission.VIEWER;
      case "editor":
        return TodoPermission.EDITOR;
      case "admin":
      case "moderator":
        return TodoPermission.MODERATOR;
      case "owner":
        return TodoPermission.OWNER;
      default:
        return TodoPermission.VIEWER;
    }
  }
  toLevel(permission: TodoPermission): TodoPermissionLevel {
    return permission as TodoPermissionLevel;
  }
  /* ════════════════════════════════════════════════════════════════════════
     CONTEXT CREATION - Create permission context from todo
     ════════════════════════════════════════════════════════════════════════ */
  createTodoPermissionContext(todo: Todo, userId?: string): TodoPermissionContext {
    const uid = userId || this.getCurrentUserId();
    const effectivePermission = this.getTodoPermission(todo, uid);
    return {
      todo,
      userId: uid,
      assigneeRoles: (todo as any).assignee_roles || {},
      effectivePermission,
    };
  }
  createTaskPermissionContext(task: Task, todo: Todo, userId?: string): TaskPermissionContext {
    const uid = userId || this.getCurrentUserId();
    const todoPermission = this.getTodoPermission(todo, uid);
    return {
      task,
      todo,
      userId: uid,
      todoPermission,
    };
  }
  createSubtaskPermissionContext(
    subtask: Subtask,
    task: Task,
    todo: Todo,
    userId?: string
  ): SubtaskPermissionContext {
    const uid = userId || this.getCurrentUserId();
    const todoPermission = this.getTodoPermission(todo, uid);
    return {
      subtask,
      task,
      todo,
      userId: uid,
      todoPermission,
    };
  }
  /* ════════════════════════════════════════════════════════════════════════
     PERMISSION CALCULATION - Single source of truth for permission logic
     ════════════════════════════════════════════════════════════════════════ */
  getTodoPermission(todo: Todo, userId: string): TodoPermission {
    if (todo.user_id === userId) {
      return TodoPermission.OWNER;
    }
    const assigneeRoles = ((todo as any).assignee_roles as Record<string, string>) || {};
    const role = assigneeRoles[userId];
    if (role) {
      return this.fromStr(role);
    }
    if (todo.visibility === "public") {
      if (this.isGlobalAdmin()) {
        return TodoPermission.MODERATOR;
      }
      return TodoPermission.VIEWER;
    }
    if (todo.visibility === "shared") {
      if (this.isGlobalAdmin()) {
        return TodoPermission.MODERATOR;
      }
      if (todo.assignees?.includes(userId)) {
        return TodoPermission.VIEWER;
      }
    }
    return TodoPermission.VIEWER;
  }
  /* ════════════════════════════════════════════════════════════════════════
     PERMISSION CHECKS - Standardized permission checks
     ════════════════════════════════════════════════════════════════════════ */
  canViewTodo(_permission: TodoPermission): boolean {
    return true;
  }
  canEditTodoFields(permission: TodoPermission): boolean {
    return [TodoPermission.MODERATOR, TodoPermission.OWNER].includes(permission);
  }
  canDeleteTodo(permission: TodoPermission): boolean {
    return permission === TodoPermission.OWNER;
  }
  canArchiveTodo(permission: TodoPermission): boolean {
    return permission === TodoPermission.OWNER;
  }
  canManageAssignees(permission: TodoPermission): boolean {
    return [TodoPermission.MODERATOR, TodoPermission.OWNER].includes(permission);
  }
  canManageGhRepo(permission: TodoPermission): boolean {
    return [TodoPermission.MODERATOR, TodoPermission.OWNER].includes(permission);
  }
  canTransferOwnership(permission: TodoPermission): boolean {
    return [TodoPermission.MODERATOR, TodoPermission.OWNER].includes(permission);
  }
  canCreateTask(permission: TodoPermission): boolean {
    if (this.isGlobalAdmin()) return true;
    return [TodoPermission.EDITOR, TodoPermission.MODERATOR, TodoPermission.OWNER].includes(
      permission
    );
  }
  canEditTask(task: Task, permission: TodoPermission, userId: string): boolean {
    if ([TodoPermission.MODERATOR, TodoPermission.OWNER].includes(permission)) {
      return true;
    }
    if (permission === TodoPermission.EDITOR) {
      return task.user_id === userId;
    }
    return false;
  }
  canDeleteTask(task: Task, permission: TodoPermission, userId: string): boolean {
    if ([TodoPermission.MODERATOR, TodoPermission.OWNER].includes(permission)) {
      return true;
    }
    if (permission === TodoPermission.EDITOR) {
      return task.user_id === userId;
    }
    return false;
  }
  canArchiveTask(task: Task, permission: TodoPermission, userId: string): boolean {
    if ([TodoPermission.MODERATOR, TodoPermission.OWNER].includes(permission)) {
      return true;
    }
    if (permission === TodoPermission.EDITOR) {
      return task.user_id === userId;
    }
    return false;
  }
  canCreateSubtask(permission: TodoPermission): boolean {
    return this.canCreateTask(permission);
  }
  canEditSubtask(subtask: Subtask, permission: TodoPermission, userId: string): boolean {
    if ([TodoPermission.MODERATOR, TodoPermission.OWNER].includes(permission)) {
      return true;
    }
    if (permission === TodoPermission.EDITOR) {
      return subtask.user_id === userId;
    }
    return false;
  }
  canDeleteSubtask(subtask: Subtask, permission: TodoPermission, userId: string): boolean {
    if ([TodoPermission.MODERATOR, TodoPermission.OWNER].includes(permission)) {
      return true;
    }
    if (permission === TodoPermission.EDITOR) {
      return subtask.user_id === userId;
    }
    return false;
  }
  canArchiveSubtask(subtask: Subtask, permission: TodoPermission, userId: string): boolean {
    if ([TodoPermission.MODERATOR, TodoPermission.OWNER].includes(permission)) {
      return true;
    }
    if (permission === TodoPermission.EDITOR) {
      return subtask.user_id === userId;
    }
    return false;
  }
  canCreateComment(permission: TodoPermission): boolean {
    return this.canCreateTask(permission);
  }
  canEditComment(comment: Comment, permission: TodoPermission, userId: string): boolean {
    if ([TodoPermission.MODERATOR, TodoPermission.OWNER].includes(permission)) {
      return true;
    }
    if (permission === TodoPermission.EDITOR) {
      return comment.user_id === userId;
    }
    return false;
  }
  canDeleteComment(comment: Comment, permission: TodoPermission, userId: string): boolean {
    if ([TodoPermission.MODERATOR, TodoPermission.OWNER].includes(permission)) {
      return true;
    }
    if (permission === TodoPermission.EDITOR) {
      return comment.user_id === userId;
    }
    return false;
  }
  canArchiveComment(comment: Comment, permission: TodoPermission, userId: string): boolean {
    if ([TodoPermission.MODERATOR, TodoPermission.OWNER].includes(permission)) {
      return true;
    }
    if (permission === TodoPermission.EDITOR) {
      return comment.user_id === userId;
    }
    return false;
  }
  /* ════════════════════════════════════════════════════════════════════════
     CONTEXT-BASED CHECKS - Use context objects for checks
     ════════════════════════════════════════════════════════════════════════ */
  checkTodoPermissions(context: TodoPermissionContext): PermissionCheckResult {
    const perm = context.effectivePermission;
    return {
      canView: this.canViewTodo(perm),
      canCreate: this.canCreateTask(perm),
      canEdit: this.canEditTodoFields(perm),
      canDelete: this.canDeleteTodo(perm),
      canArchive: this.canArchiveTodo(perm),
      canManageAssignees: this.canManageAssignees(perm),
      permissionLevel: this.toLevel(perm),
    };
  }
  checkTaskPermissions(context: TaskPermissionContext): PermissionCheckResult {
    const perm = context.todoPermission;
    return {
      canView: this.canViewTodo(perm),
      canCreate: this.canCreateTask(perm),
      canEdit: this.canEditTask(context.task, perm, context.userId),
      canDelete: this.canDeleteTask(context.task, perm, context.userId),
      canArchive: this.canArchiveTask(context.task, perm, context.userId),
      canManageAssignees: this.canManageAssignees(perm),
      permissionLevel: this.toLevel(perm),
    };
  }
  checkSubtaskPermissions(context: SubtaskPermissionContext): PermissionCheckResult {
    const perm = context.todoPermission;
    return {
      canView: this.canViewTodo(perm),
      canCreate: this.canCreateSubtask(perm),
      canEdit: this.canEditSubtask(context.subtask, perm, context.userId),
      canDelete: this.canDeleteSubtask(context.subtask, perm, context.userId),
      canArchive: this.canArchiveSubtask(context.subtask, perm, context.userId),
      canManageAssignees: this.canManageAssignees(perm),
      permissionLevel: this.toLevel(perm),
    };
  }
  /* ════════════════════════════════════════════════════════════════════════
     ASYNC PERMISSION LOOKUP - For cases requiring server-side lookup
     ════════════════════════════════════════════════════════════════════════ */
  async getTodoPermissionsAsync(
    todoId: string,
    visibility: string,
    token: string
  ): Promise<Record<string, string>> {
    try {
      const result: any = await firstValueFrom(
        (this.api as any).invokeCommand("get_todo_permissions", {
          todo_id: todoId,
          visibility: visibility || "private",
          token: token || "",
        })
      );
      return result?.assignee_roles || result || {};
    } catch (e) {
      return {};
    }
  }
  getDefaultRole(): TodoPermission {
    return TodoPermission.VIEWER;
  }
  /* ════════════════════════════════════════════════════════════════════════
     UTILITY METHODS
     ════════════════════════════════════════════════════════════════════════ */
  isOwner(todo: Todo, userId?: string): boolean {
    const uid = userId || this.getCurrentUserId();
    return todo.user_id === uid;
  }
  isEditor(todo: Todo, userId?: string): boolean {
    const perm = this.getTodoPermission(todo, userId || this.getCurrentUserId());
    return perm === TodoPermission.EDITOR;
  }
  isModerator(todo: Todo, userId?: string): boolean {
    const perm = this.getTodoPermission(todo, userId || this.getCurrentUserId());
    return perm === TodoPermission.MODERATOR;
  }
  clearPermissionCache(): void {
    this.permissionCache.set(new Map());
  }
}
